core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -946,7 +946,7 @@ public SqlNode field(int ordinal) {
                 && neededAlias != null
                 && (aliases.size() != 1 || !aliases.containsKey(neededAlias))) {
           newContext =
-              aliasContext(ImmutableMap.of(neededAlias, rel.getRowType()),
+              aliasContext(ImmutableMap.of(neededAlias, rel.getInput(0).getRowType()),
```

Comment:
Although this code is only ever reached for instances of SingleRel, this doesn't seem like a very elegant solution.  Any pointers or edits would be appreciated.